src/librustc_data_structures/indexed_vec.rs

```diff
@@ -75,23 +92,69 @@ macro_rules! newtype_index {
 
     // Base case, user-defined constants (if any) have already been defined
     (@derives      [$($derives:ident,)*]
-     @pub          [$($pub:tt)*]
      @type         [$type:ident]
      @max          [$max:expr]
+     @vis          [$v:vis]
      @debug_format [$debug_format:tt]) => (
         #[derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, $($derives),*)]
-        pub struct $type($($pub)* u32);
+        $v struct $type {
+            private: ::std::num::NonZeroU32
+        }
+
+        impl $type {
+            $v const MAX_AS_U32: u32 = $max;
+
+            $v const MAX: $type = unsafe { $type::from_u32_unchecked($max) };
+
+            #[inline]
+            $v fn from_usize(value: usize) -> Self {
+                assert!(value <= ($max as usize));
+                unsafe {
+                    $type::from_u32_unchecked(value as u32)
+                }
+            }
+
+            #[inline]
+            $v fn from_u32(value: u32) -> Self {
+                assert!(value <= $max);
+                unsafe {
+                    $type::from_u32_unchecked(value)
+                }
+            }
+
+            #[inline]
+            $v const unsafe fn from_u32_unchecked(value: u32) -> Self {
+                $type { private: ::std::num::NonZeroU32::new_unchecked(value + 1) }
+            }
+
+            /// Extract value of this index as an integer.
+            #[inline]
+            $v fn index(self) -> usize {
+                self.as_usize()
+            }
+
+            /// Extract value of this index as a usize.
+            #[inline]
+            $v fn as_u32(self) -> u32 {
+                self.private.get() - 1
```

Comment:
I didn't realize everything was incremented by `1`! This really needs benchmarking IMO.

Reply:
PS you should add `NonMaxU32` already ;)

Reply:
I've been blocking it on const-generics, if you want to add some compiler-private lang items, feel free to go ahead and do it, the part that controls the valid range is *tiny*:
https://github.com/rust-lang/rust/blob/d67ba90dab243d2b80324e29df2f06ba13c97607/src/librustc/ty/layout.rs#L764-L775

(for non-max things you'd use `0..=*scalar.valid_range.end()-1` instead of `1..=*scalar.valid_range.end()`)